src/main/java/uk/gov/ons/ctp/response/collection/exercise/validation/ValidateSampleUnits.java

```diff
@@ -68,75 +58,110 @@
+
+            List<String> classifierTypes =
+                requestSurveyClassifiers(sampleUnit.getSampleUnitGroup().getCollectionExercise());
 
-      collections.forEach(
-          (exercise, groups) -> {
             try {
-              addCollectionInstrumentIds(exercise, groups);
-              transitionCollectionExercise(exercise);
+              UUID collectionInstrumentId =
+                  requestCollectionInstrumentId(
+                      classifierTypes,
+                      sampleUnit,
+                      sampleUnit
+                          .getSampleUnitGroup()
+                          .getCollectionExercise()
+                          .getSurveyId()
+                          .toString());
+              sampleUnit.setCollectionInstrumentId(collectionInstrumentId);
+            } catch (HttpClientErrorException e) {
+              if (e.getStatusCode() != HttpStatus.NOT_FOUND) {
+                log.with("sample_unit", sampleUnit)
+                    .with("status_code", e.getStatusCode())
+                    .error("Unexpected HTTP response code from collection instrument service");
+                throw e; // Re-throw anything that's not a 404 so that we retry
+              }
+            }
+
+            if (sampleUnit.getSampleUnitType() == SampleUnitDTO.SampleUnitType.B) {
+              try {
+                PartyDTO party =
+                    partySvcClient.requestParty(
+                        sampleUnit.getSampleUnitType(), sampleUnit.getSampleUnitRef());
+                sampleUnit.setPartyId(UUID.fromString(party.getId()));
+              } catch (HttpClientErrorException e) {
+                if (e.getStatusCode() != HttpStatus.NOT_FOUND) {
+                  log.with("sample_unit", sampleUnit)
+                      .with("status_code", e.getStatusCode())
+                      .error("Unexpected HTTP response code from party service");
+                  throw e; // Re-throw anything that's not a 404 so that we retry
+                }
+              }
+            }
+
+            ExerciseSampleUnitGroup sampleUnitGroup = sampleUnit.getSampleUnitGroup();
+            sampleUnitGroup.setModifiedDateTime(new Timestamp(new Date().getTime()));
+
+            SampleUnitGroupEvent event = SampleUnitGroupEvent.VALIDATE;
+            if (!isSampleUnitValid(sampleUnit)) {
+              event = SampleUnitGroupEvent.INVALIDATE;
+            }
+
+            try {
+              sampleUnitGroup.setStateFK(
+                  sampleUnitGroupState.transition(
+                      sampleUnit.getSampleUnitGroup().getStateFK(), event));
             } catch (CTPException e) {
-              log.error(
-                  "Error validating collection exercise, collectionExerciseId: {}",
-                  exercise.getId());
-              log.error("Stack trace: {}", e);
+              // Ignored because this should never happen - we know the transition is valid
```

Comment:
Is it safe to ever do this? Shouldn't it at least log? _Things that should never happen_ tend to happen more often than you think :)

Reply:
At the very least it'll mean it'll start screaming if we mess with the state machine and introduce a regression

Reply:
it's the state machine code that's wrong - it shouldn't throw a checked exception. I suppose I could re-throw the exception as a runtime exception

Reply:
We just don't want exceptions disappearing into the ether if they occur (even if they're really not supposed to). Everything should be handled even if that's just "something that shouldn't have happened has happened".